stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -62,35 +81,96 @@
-        (map-set tally tally-key new-total)
-        (map-set used-aggregate-public-keys key {reward-cycle: reward-cycle, round: round})
-        (update-last-round reward-cycle round)
-        (print { 
-            event: "voted", 
-            signer: tx-sender, 
-            reward-cycle: reward-cycle, 
-            round: round, 
-            key: key, 
-            new-total: new-total })
-        (ok true)))
+(define-private (sum-weights (signer { signer: principal, weight: uint }) (acc uint))
+    (+ acc (get weight signer)))
+
+(define-private (get-total-weight (reward-cycle uint))
+    (match (map-get? cycle-total-weight reward-cycle)
+        total (ok total)
+        (let ((signers (unwrap! (contract-call? .signers get-signers reward-cycle) (err ERR_FAILED_TO_RETRIEVE_SIGNERS)))
+                (total (fold sum-weights signers u0)))
+            (map-set cycle-total-weight reward-cycle total)
+            (ok total))))
 
 (define-private (update-last-round (reward-cycle uint) (round uint))
-    (match (map-get? rounds reward-cycle)
-        last-round (and (> round last-round) (map-set rounds reward-cycle round))
-        (map-set rounds reward-cycle round)))
+    (ok (match (map-get? rounds reward-cycle)
+        last-round (begin
+            (asserts! (<= round (+ last-round u1)) (err ERR_INVALID_ROUND))
+            (and (> round last-round) (map-set rounds reward-cycle round)))
+        (map-set rounds reward-cycle round))))
+
+;; Signer vote for the aggregate public key of the next reward cycle
+;;  Each signer votes for the aggregate public key for the next reward cycle.
+;;  This vote must happen after the list of signers has been set by the node,
+;;  which occurs in the first block of the prepare phase. The vote is concluded
+;;  when the threshold of `threshold-consensus / 1000` is reached for a
+;;  specific aggregate public key. The vote is weighted by the amount of
+;;  reward slots that the signer controls in the next reward cycle. The vote
+;;  may require multiple rounds to reach consensus, but once consensus is
+;;  reached, later rounds will be ignored.
+;;
+;;  Arguments:
+;;   * signer-index: the index of the calling signer in the signer set (from
+;;     `get-signers` in the .signers contract)
+;;   * key: the aggregate public key that this vote is in support of
+;;   * round: the voting round for which this vote is intended
+;;   * reward-cycle: the reward cycle for which this vote is intended
+;;  Returns:
+;;   * `(ok true)` if the vote was successful
+;;   * `(err <code>)` if the vote was not successful (see errors above)
+(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint) (reward-cycle uint))
+    (let ((tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
+            ;; vote by signer weight
+            (signer-weight (try! (get-current-signer-weight signer-index)))
```

Comment:
I think there's potentially a logic bug here.  This function takes a `reward-cycle` as an argument, but it uses `get-current-signer-weight` instead of `get-signer-weight`.  This implicitly assumes `reward-cycle` is the same as the upcoming reward cycle, but this is never checked.

Reply:
Generally speaking, I'd prefer it if Clarity code didn't use globals except in the top-level calling context (and even then, only sparingly).  The factoring I'd like to see is as follows:

```clarity
(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint) (reward-cycle uint))
   (begin
      (asserts! (is-eq reward-cycle (+ u1 (burn-height-to-reward-cycle burn-block-height)))
         (err ERR_OUT_OF_VOTING_WINDOW))
      (inner-vote-for-aggregate-public-key signer-index key round reward-cycle)))

(define-private (inner-vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint) (reward-cycle uint))
   ;; etc. etc., but use `get-signer-weight` instead of `get-current-signer-weight`
```

Then, it's straightforward to unit-test and property-test `inner-vote-for-aggregate-public-key` without having to mock the global state.

Reply:
I fixed that original issue in f596c0bf5b0c463859b2a25d0322f40a6efd8c0d.

We talked about the fact that we may at some point want to allow votes for cycle N to happen during cycle N, in which case, we won't want that new error check.